doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,326 @@
+            'access' => 'readwrite',
+            'setaction' => qq(
+                if (!set_property_query(ctx, p->data) || !set_digest(ctx))
+                    return 0;
+            ),
+        ),
+
+        'KDF_PARAM_SIZE' => (
+            'type' => 'integer',
+            'ctype' => 'size_t',
+            'access' => 'readonly',
+            'field' => "SIZE_MAX",
+        ),
+    );
+);
+-}
+/* End of generated code */
+
+THe top level attributes are:
+
+- "name" is the name the functions will derive from e.g. "kdf_scrypt" to this will be appended _[gs]et[_ctx]_params
+- "functions" is the functions to generate.  By default both setters and getters but either can be omitted.
+- "prologue" defines some introductory code emited in the generated functions.  Function arguments are:
+  `void *vctx, OSSL_PARAM params[]` and this can be used to specialise the void pointer or declare locals.
+- "epilogue" defines some post decode code emited in the generated function
+- "params" defines the parameters both gettable and settable
+
+Within the "params" the fields specify each parameter by label.
+
+Each parameter is then specialised with attibutes:
+
+- "type" is the OSSL_PARAM type
+- "ctype" is the underlying C type (e.g. for an integer parameter size_t could be the C type)
+- "access" is readwrite, readonly or writeonly.  This determines if the parameter is a settable, gettable or both
+- "field" is an accessor to the field itself
+- "sanitycheck" is a validation check for the parameter.  If present, code will be generated `if (!(sanitycheck)) return 0;`
+  The local variable `var` will contain the C value if specified.
+- "setaction" is C code to execute when the parameter is being set.  It will define an OSSL_PARAM pointer p to set.
+- "code" set to "no" skips code generation for this parameter, it defaults to "yes" which generates handlers.  This is useful
+  when a parameter is duplicated with differenting types (e.g. utf8 string and integer).
+- "published" set to "yes" includes the parameter in the gettable/settable lists.  Set to "no" and it isn't included (but will
+  still be processed).  It defaults to "yes".
+
+- Flags include:
+  - nostatic: do not make the function static
+  - nocode: do not generate code for this parameter
+    - This allows, e.g., two different types for a parameter (int & string)
+  - unpublished: do not generate this parameter in the gettable/settable list
+    - "engine" is the only one like this
+  - readonly: create a getter but not a setter
+  - writeonly: create a setting but not a getter
+
+The idea is that the gettable and get functions will be simultaneously generated
+along with fast decoder to look up parameter names quickly.
+
+The getter and setter functions will be pre-populated with some local variable:
+
+    OSSL_PARAM *p;          /* The matching parameter */
+    type val;               /* The value of the parameter after a get/set call (for C types) */
+
+
+A worked example for scrypt:
+
```

Comment:
I would like to see the actual example.

Reply:
Ah, never mind, it's above.

Reply:
That is the example, output C code is further down (although unlikely to be all that close to the final output).

Until time is invested into implementing this, there won't & cannot be an actual example.